modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+                          .blockOn(blocker) {
+                            F.bracketCase(F.delay(producer.beginTransaction())) { _ =>
+                              committable.records
+                                .traverse(
+                                  KafkaProducer
+                                    .produceRecord(keySerializer, valueSerializer, producer)
+                                )
+                                .flatMap(_.sequence)
+                                .flatTap { _ =>
+                                  F.delay {
+                                    producer.sendOffsetsToTransaction(
+                                      batch.offsets.asJava,
+                                      groupId
+                                    )
+                                  }
+                                }
+                            } {
+                              case (_, ExitCase.Completed) =>
+                                F.delay(producer.commitTransaction())
+                              case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
+                                F.delay(producer.abortTransaction())
                             }
                           }
-                      } {
-                        case (_, ExitCase.Completed) =>
-                          F.delay(producer.commitTransaction())
-                        case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
-                          F.delay(producer.abortTransaction())
                       }
-                    }.flatten
+                    }
                   }
-                }
-              }
 
-              override def toString: String =
-                "TransactionalKafkaProducer$" + System.identityHashCode(this)
+                override def toString: String =
+                  "TransactionalKafkaProducer$" + System.identityHashCode(this)
+              }
             }
           }
+        }
+      }
+
+    private def init(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
+      context.blockOn(blocker)(F.delay(producer.initTransactions()))
+
+    private def close(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
+      context.blockOn(blocker)(F.delay(producer.close(settings.closeTimeout.asJava)))
+
+    /**
+      * Construct a [[TransactionalKafkaProducer]] which uses a constant ID
+      * for all transactions.
+      *
+      * NOTE: This mode only guarantees exactly-once processing if you have a
+      * single instance in the consumer group. True exactly-once processing is
+      * only guaranteed by running separate producers with unique transactional
+      * IDs per topic/partition processed by the group.
+      */
+    def withConstantId: Resource[F, TransactionalKafkaProducer[F, K, V]] =
+      resource { blocker =>
+        Resource
+          .make(settings.createProducer)(close(blocker))
+          .evalTap(init(blocker))
+          .map { producer => records =>
```

Comment:
This is the other spot where removing `G[_]` pays off. Without the method-level type parameter we can pass the grouping logic around as a simple function. If we keep the `G[_]`, we have to introduce a new `trait` with a generic method and construct that here instead.